src/content/CloudBoilerplateNet/Services/ReactiveCacheManager.cs

```diff
@@ -0,0 +1,273 @@
+{
+    public class ReactiveCacheManager : ICacheManager
+    {
+        #region "Fields"
+
+        private readonly IDependentTypesResolver _dependentTypesResolver;
+        private readonly object _dummyEntryCreationLock = new object();
+        private readonly object _entryCreationLock = new object();
+        private bool _disposed;
+
+        #endregion
+
+        #region "Properties"
+
+        public int CacheExpirySeconds
+        {
+            get;
+            set;
+        }
+
+        public IMemoryCache MemoryCache { get; }
+
+        protected List<string> InvalidatingOperations => new List<string>
+        {
+            "upsert",
+            "publish",
+            "restore_publish",
+            "unpublish",
+            "archive",
+            "restore"
+        };
+
+        #endregion
+
+        #region "Constructors"
+
+        public ReactiveCacheManager(IOptions<ProjectOptions> projectOptions, IMemoryCache memoryCache, IDependentTypesResolver relatedTypesResolver, IWebhookListener webhookListener)
+        {
+            CacheExpirySeconds = projectOptions?.Value?.CacheTimeoutSeconds ?? throw new ArgumentNullException(nameof(projectOptions));
+            MemoryCache = memoryCache ?? throw new ArgumentNullException(nameof(memoryCache));
+            _dependentTypesResolver = relatedTypesResolver ?? throw new ArgumentNullException(nameof(relatedTypesResolver));
+
+            WebhookObservableFactory
+                .GetObservable(webhookListener, nameof(webhookListener.WebhookNotification))
+                .Where(args => InvalidatingOperations.Any(operation => operation.Equals(args.Operation, StringComparison.Ordinal)))
+                .Throttle(TimeSpan.FromSeconds(1))
+                .DistinctUntilChanged()
+                .Subscribe((args) => InvalidateEntry(args.IdentifierSet));
+        }
+
+        #endregion
+
+        #region "Public methods"
+
+        /// <summary>
+        /// Gets an existing cache entry or creates one using the supplied <paramref name="valueFactory"/>.
+        /// </summary>
+        /// <typeparam name="T">Type of the cache entry value.</typeparam>
+        /// <param name="identifierTokens">String tokens that form a unique identifier of the entry.</param>
+        /// <param name="valueFactory">Method to create the entry.</param>
+        /// <param name="dependencyListFactory">Method to get a collection of identifiers of entries that the current entry depends upon.</param>
+        /// <returns>The cache entry value, either cached or obtained through the <paramref name="valueFactory"/>.</returns>
+        public async Task<T> GetOrCreateAsync<T>(IEnumerable<string> identifierTokens, Func<Task<T>> valueFactory, Func<T, IEnumerable<IdentifierSet>> dependencyListFactory, bool awaitCacheEntryCreation = true)
+        {
+            // Check existence of the cache entry.
+            if (!MemoryCache.TryGetValue(StringHelpers.Join(identifierTokens), out T entry))
```

Comment:
tune the locking as we've discussed - we want to prevent the `valueFactory();` to be called if not necessary...aka double-checked locking :)

Reply:
you may find `awaitCacheEntryCreation` unnecessary and leave only the synchronous branch of the code ;)

Reply:
At the first glance, the omission of the `lock` in the outer `GetOrCreateAsync` might seem as a mistake to one (including myself). But, think of the pros and cons of the outer lock.

To me, the most important thing is to prevent the concurrent creation of the cache entry and the dummy entries. This might lead to unpredictable MemoryCache behavior (broken CancellationToken references, duplicate content items in the cache). This misbehavior is prevented by the locking inside of `CreateEntry`.

Now, what happens when we lock the whole outer execution path in `GetOrCreateAsync`? Yes, we'd prevent a few duplicate Delivery API requests during the time of creating the cache entry of the first of the potentially-duplicate requests. But for what price? All of the DeliveryClient queries of all the threads would start to stack up in a sequential (and possibly blocking) manner. GetOrCreateAsync would become a bottleneck IMO. 

I should emphasize that the locking:

* is here to help in a rare situation when multiple concurrent threads issue an identically-parametrized Delivery call
* and it only helps during the short period when the cache entry of the first-coming call is under construction

With that in mind, wouldn't it be an overkill to lock everything into a sequential processing?

Apart from that, it is only possible to await a `Task` using the `SemaphoreSlim` object, not inside of `lock`. Also, I'm not sure what would be the consequences of using semaphores.

What do you think?

Reply:
1) Most importantnly, I think that you should lock based on cache keys as outlined [here](http://www.java2s.com/Code/CSharp/Development-Class/MemoryCacheLock.htm).

2) Yes, you might need to use `SemaphoreSlim`

3) And, I will repeat myself and say that I wouldn't be doing `Task.Run(() => CreateEntry())` unless I had a very good reason. The biggest portion of computing time will be consumed by data retrieval which you need to do anyway. Compared to that, creation of a cache item seems like an insignificant overhead...

Reply:
When tested, [getting data from the US KC datacenter to the app in EU](https://github.com/Kentico/cloud-boilerplate-net/blob/webhook-cache-invalidation/src/content/CloudBoilerplateNet/Services/ReactiveCacheManager.cs#L83) took  **~150ms**.

A [call to CreateEntry](https://github.com/Kentico/cloud-boilerplate-net/blob/webhook-cache-invalidation/src/content/CloudBoilerplateNet/Services/ReactiveCacheManager.cs#L87) for that data took **582ms** due to necessary usage of reflection during the dependency creation process (on a very fast machine).

Therefore, I've tuned things up but left the Task.Run() in place. It can always be switched on/off via [ProjectOptions.CreateCacheEntriesInBackground](https://github.com/Kentico/cloud-boilerplate-net/blob/webhook-cache-invalidation/src/content/CloudBoilerplateNet/ProjectOptions.cs#L13).